### PR TITLE
Add array_equal to PyOpenCLArrayContext.np

### PR DESCRIPTION
**Pull request by majosm**
Used by #46.

I'm not sure if there's a way to also make this work for `PytatoPyOpenCLArrayContext`? I ran into trouble with `DataWrapper`'s `__eq__` when I tried it.

### CODE REVIEW

**Comment by kaushikcfd** [comment]
> I ran into trouble with DataWrapper's __eq__ when I tried it.

For pytato arrays `Array.__eq__` is not compatiable with numpy. Maybe you want [pt.equal](https://documen.tician.de/pytato/array.html?highlight=equal#pytato.equal). But we don't have `(all|any)` implemented for pytato arrays as of yet.

File: arraycontext/impl/pyopencl/fake_numpy.py [hunk 1]
```diff
@@ -208,6 +210,31 @@ def all(self, a):
...
+        elif not is_array_container(a):
+            same_shape = a.shape == b.shape
+            if same_shape:
+                result = (a == b).all()
+            else:
+                result = as_scalar(False)
+        else:
+            queue = self._array_context.queue
+            result = rec_multimap_reduce_array_container(
+                    partial(reduce, partial(cl_array.minimum, queue=queue)),
```
Comment (inducer) [inline arraycontext/impl/pyopencl/fake_numpy.py:230]: Is `minimum` documented to do the right thing applied to `bool` arrays?

File: arraycontext/impl/pyopencl/fake_numpy.py [hunk 2]
```diff
@@ -208,6 +210,31 @@ def all(self, a):
             result = result.get()[()]
         return result
 
+    def array_equal(self, a, b):
+        def as_scalar(bool_value):
```
Comment (inducer) [inline arraycontext/impl/pyopencl/fake_numpy.py:214]: This seems to unnecessarily round-trip to the device. Why not just return the (known) `bool` result?

**Review by inducer — COMMENTED** [review]
Thanks! A few notes below.

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

File: arraycontext/impl/pyopencl/fake_numpy.py [hunk 3]
```diff
@@ -208,6 +210,31 @@ def all(self, a):
             result = result.get()[()]
         return result
 
+    def array_equal(self, a, b):
```
Comment (inducer) [inline arraycontext/impl/pyopencl/fake_numpy.py:213]: @kaushikcfd What do you think this should return? Device scalar or "just plain `bool`"?

Comment (alexfikl) [reply arraycontext/impl/pyopencl/fake_numpy.py:213]: For what it's worth, the other reductions (e.g. `any`) seem to return a device scalar if given arrays and a host scalar if given host scalars.

Comment (kaushikcfd) [reply arraycontext/impl/pyopencl/fake_numpy.py:213]: I agree with @alexfikl, we should probably avoid implicit freeze-ing as much as possible.

File: arraycontext/impl/pyopencl/fake_numpy.py (hunk 1 above)
Comment (majosm) [reply arraycontext/impl/pyopencl/fake_numpy.py:230]: I separated the device-to-host conversion from the recursion step, so I think this should only be operating on `int8` arrays now.

File: arraycontext/impl/pyopencl/fake_numpy.py [hunk 4]
```diff
@@ -208,6 +211,31 @@ def all(self, a):
...
+        def rec_equal(x, y):
+            if type(x) != type(y):
+                return as_device_scalar(False)
+            elif not is_array_container(x):
+                return as_device_scalar(x.shape == y.shape) and (x == y).all()
+            else:
+                queue = self._array_context.queue
+                return multimap_reduce_array_container(
+                    partial(reduce, partial(cl_array.minimum, queue=queue)),
+                    rec_equal, x, y)
```
Comment (inducer) [inline arraycontext/impl/pyopencl/fake_numpy.py:231]:
```suggestion
                    # reduce_func=
                    partial(reduce, partial(cl_array.minimum, queue=queue)),
                    # map_func=
                    rec_equal,
                    x, y)
```

**Review by inducer — APPROVED** [review]

**Comment by inducer** [comment]
Thx!

**Pull request merged**